### ISSUE TITLE
Bukkit dependency should be updated to spigot-api

### ISSUE THREAD

**Issue by haveric** [MilkBowl/VaultAPI#116]
The bukkit dependency is not being provided in newer versions of spigot releases (even aikar stopped providing them past 1.16.4). Due to this, there is no override-able bukkit version and projects will try to compile for 1.13.1 coming from Vault instead of newer versions using the spigot-api.

**Comment by Geolykt** [comment]
You can usually exclude transitive dependencies in your poms, which would solve this issue

**Comment by haveric** [comment]
Thanks. I wasn't sure that was going to work initially as I had some other dependency issues, but I was able to do so. For anyone else having this issue with gradle:
```
compileOnly("com.github.MilkBowl:VaultAPI:1.7") {
    exclude group: 'org.bukkit', module: 'bukkit'
}
```

It still probably makes sense to update to spigot-api on the VaultAPI side at some point though, but I'll leave that for the devs to decide.

**Comment by A248** [comment]
Changing Vault's dependency declaration won't solve the underlying issue. There are various ways to depend on versions of the Bukkit API. You can depend on the Paper API, the Spigot API, the Tuinity API, the APIs for other forks, etc.

The problem is that Vault ought to declare the scope correctly. In Maven, this would be `provided`. In Gradle, it would be `compileOnly`.

Consumers can then choose which Bukkit API dependency they wish to build with.

**Comment by A248** [comment]
Seemingly already solved: https://github.com/SolarMC-Dev/VaultAPI/commit/c17036183c0c32f89db12b1f1540503859b697a1

**Comment by A248** [comment]
Seeming already solved: https://github.com/MilkBowl/VaultAPI/commit/c17036183c0c32f89db12b1f1540503859b697a1

**Comment by haveric** [comment]
> Changing Vault's dependency declaration won't solve the underlying issue.

Yes actually it would. There is no bukkit dependency being provided in newer versions, making spigot the main dependency now. You can ignore all of the others (Paper, Tuinity, etc.) because they all depend on spigot.

**Comment by A248** [comment]
> making spigot the main dependency now. You can ignore all of the others (Paper, Tuinity, etc.) because they all depend on spigot.

This isn't true. There is no "main dependency". The Paper API artifact does not depend on the Spigot API artifact - it is a separate API and has way more features than the Spigot API, even in the same classes.

Neither is it the case for other forks. They each have their own API artifacts which do not depend on other versions of the Bukkit API.

**Comment by haveric** [comment]
There is no [Bukkit API](https://hub.spigotmc.org/nexus/content/groups/public/org/bukkit/bukkit/) being provided for current versions, which would mean that you can no longer depend upon it. While I agree that internally the spigot team is building spigot off of bukkit, unless there is a bukkit API being provided separately, it really doesn't matter. At some point, the bukkit dependency will need to be updated to 1.16+ within Vault and there won't be an option other than spigot. That's all I'm trying to get at here.

**Comment by A248** [comment]
I don't know what you're trying to say. Do you understand that there are separate API artifacts which all provide the bukkit API?

I don't depend on Spigot myself and my colleagues don't deploy it. Many others are in the same position as me, and there is no reason for us to depend on the `org.spigotmc:spigot-api` artifact. So, as I've said before, the solution is for Vault to declare the dependency provided, so API consumers are not forced to depend on whatever bukkit API artifact (whether that is org.spigot:spigot-api or io.papermc.paper:paper-api) is chosen.

**Comment by haveric** [comment]
I do understand that there are other API artifacts. My understanding is that each of the others pulls in everything from Spigot, and the spigot-api was the replacement for the bukkit API going forward (as of 1.16), but that may be a flawed assumption. If that assumption is valid, then this project should be able to seamlessly switch to the spigot-api and nobody would notice a difference other than less development issues such as what I ran into (and was able to workaround). If that is not how other projects are handling things, I'm not sure how this plugin would be able to handle the bukkit dependency going forward, or for that matter any bukkit/spigot/whatever plugin would without providing different jars for each as they all rely on the spigot provided api as the starting point. I suppose an alternative would be for someone else to provide the 1.16+ API jars, but unless that happens, there will be dependency issues at some point down the road.

**Comment by A248** [comment]
I'm not sure you quite understand the issue here. Either that, or you're not well articulating what you think the problem is. Your wording is very vague so it's hard for me to pinpoint why exactly you think Vault should switch to `org.spigotmc:spigot-api`. Let me layout what I've been referring to and hopefully offer better clarity.

The issue, as I understand it, is that Vault has a transitive dependency on the `org.bukkit:bukkit` artifact. Traditional build system dependency resolution, whereby multiple versions of the same artifact are consolidated to a single version, is not effective in this case, because `org.bukkit:bukkit`, `org.spigotmc:spigot-api`, and `io.papermc.paper:paper-api` are all different artifacts. They have different groupId/artifactId combinations, to use Maven terms.

Accordingly, dependents of the Vault API see both `org.bukkit:bukkit` and the other bukkit API artifact they have chosen. The other bukkit API artifact may be `io.papermc.paper:paper-api` , `org.spigotmc:spigot-api`, or any other artifact which provides the group of classes in the org.bukkit namespace constituting the "Bukkit API." When dependencies with identically-named classes are on the classpath, one of the classes is chosen. Accordingly, in the case that `org.bukkit:bukkit` is prioritized by the build system over the other bukkit API artifact, the dependent of the Vault API inadvertently compiles against `org.bukkit:bukkit`. This is problematic for obvious reasons.

The proper solution to this issue is for Vault to declare the dependency as non-transitive. This solution has already been applied to the main branch. Unfortunately, this solution is not yet available to dependents, because the version has not been incremented in the build file. Therefore, the artifact served by Jitpack is still the original 1.7 artifact, because of Jitpack's immutable artifacts policy for release versions.

What you appear to be proposing is to change the dependency from `org.bukkit:bukkit` to `org.spigotmc:spigot-api`. In terms of alleviating dependency conflicts in dependents, this does not fix the issue. It merely changes which artifact happens to be declared as a transitive dependency. Dependents who do not use the spigot-api will still have the exact same described issue with dependency resolution.

> My understanding is that each of the others pulls in everything from Spigot, and the spigot-api was the replacement for the bukkit API going forward (as of 1.16), but that may be a flawed assumption
> All of the other APIs should be providing a full Spigot API at this point

I do not understand what you intend by "pulls in everything from Spigot" or "providing a full Spigot API at this point." These terms could relate to dependency resolution or to binary compatibility.

If you are referring to binary compatibility, then it does not matter from an API dependent's perspective which dependency is declared by the Vault API. Whether the plugin provision of Vault is compatible with specific server software is relevant for server administrators and would be best directed to the [Vault](https://github.com/MilkBowl/Vault) repository. Changing the declared artifact which provides the bukkit API may assist the Vault team in maintaining compatibility. Nevertheless, it is entirely possible for the Vault team to internally depend on an artifact providing the bukkit API other than `org.spigotmc:spigot-api`.

**Comment by haveric** [comment]
Look, I never said that it had to be changed to the spigot-api, and was only offering it as a suggestion. I was looking at it from the perspective of if you were to drop legacy support sometime in the future and still depend on bukkit, your two choices would be to build your own bukkit dependency or depend on spigot-api (or I guess one of the others) due to the dependency you are using no longer being provided from the source for newer versions.

I was also coming from the perspective that any developer that has updated to provide support for the 1.17 api has no choice now, but to rely on the spigot-api unless they build the bukkit api themselves, which I assume basically nobody will, making spigot-api the next best and recommended option from the spigot devs themselves. If there's some alternative that I'm not aware of, I would love to know what it is, but seeing as every other plugin would then be relying on the spigot-api, I thought it could prevent time troubleshooting for more people by Vault updating as well.

With that said, my plugins have already implemented the solution I mentioned above and it sounds like the fix you've mentioned already is a fine solution, but due to the immutable artifacts on jitpack, the solution hasn't really taken effect. Would it make sense then to release a minor version update in order to get the fix properly deployed, especially since it was implemented a while ago? 

I know it seems like a small thing to consider, but I know when I ran into the issue, it was not obvious how to fix it and caused me plenty of frustration in depending on a plugin that I really did not want to have to drop support for, which is why I initially came here for support in the first place.

**Comment by Sleaker** [comment]
https://github.com/MilkBowl/VaultAPI/issues/116#issuecomment-859937326

This is the correct way to handle transitive dependency problems due to changes in how build systems have changed over time.


> 
> 
> Seeming already solved: [c170361](https://github.com/MilkBowl/VaultAPI/commit/c17036183c0c32f89db12b1f1540503859b697a1)

Correct. this was already resolved.
